specs/trip_dispatch_spec.rb

```diff
@@ -88,5 +97,231 @@
+
+    it "selects a driver who doesn't have any trips with an end-time of nil" do
+      next_available_driver = @dispatcher.drivers.find do |driver|
+        driver.status == :AVAILABLE
+      end
+      nil_end_time = false
+      next_available_driver.trips.each do |trip|
+        nil_end_time = trip.end_time.nil?
+      end
+
+      nil_end_time.must_equal false
+    end
+
+    it "assigns the first new driver (hasn't completed trips yet) then available drivers whose most recent trip is the earliest in time" do
+
+      first_driver = @dispatcher.find_driver(100)
+
+      second_driver = @dispatcher.find_driver(14)
+
+      third_driver = @dispatcher.find_driver(27)
+
+      fourth_driver = @dispatcher.find_driver(6)
+
+      fifth_driver = @dispatcher.find_driver(87)
+
+      sixth_driver = @dispatcher.find_driver(75)
+
+      @result.driver.id.must_equal first_driver.id
+
+      result = @dispatcher.request_trip(6)
+      result.driver.id.must_equal second_driver.id
+      result = @dispatcher.request_trip(7)
+      result.driver.id.must_equal third_driver.id
+      result = @dispatcher.request_trip(8)
+      result.driver.id.must_equal fourth_driver.id
+      result = @dispatcher.request_trip(9)
+      result.driver.id.must_equal fifth_driver.id
+      result = @dispatcher.request_trip(10)
+      result.driver.id.must_equal sixth_driver.id
+
+    end
+
+    it "sets the selected driver's status to unavailable" do
+      @result.driver.status.must_equal :UNAVAILABLE
+    end
+
+    it "raises an error if there are no available drivers" do
+      drivers = @dispatcher.drivers
+      available_drivers = drivers.select do |driver|
+        driver.status == :AVAILABLE && !driver.trips.empty?
+      end
+      available_count = available_drivers.length
+
+      available_count.times do
+        trip = @dispatcher.request_trip(2)
+      end
+
+      trips_length = @dispatcher.trips.length
+
+      result = proc { @dispatcher.request_trip(2) }
+
+      updated_trips_length = @dispatcher.trips.length
+
+      result.must_raise StandardError
+      updated_trips_length.must_equal trips_length
+
+    end
+
+    it "uses the current time for the start time" do
+      current_time = Time.now
+      @result.start_time.to_a.must_equal current_time.to_a
+    end
+
+    it "assigns end time, cost, and rating to nil" do
+      @result.end_time.must_be_nil
+      @cost.must_be_nil
+      @rating.must_be_nil
+    end
+
+    it "updates the trips list" do
+      @dispatcher.trips.must_include @result
+    end
+
+    it "updates the trip list for the driver" do
+      result = @result.driver.trips
+      result.must_include @result
+    end
+
+    it "updates the trip list for the passenger" do
+      result = @result.passenger.trips
+      result.must_include @result
+    end
+
+    describe "pending trips excluded from calculations involving end_time, cost, and rating" do
+      before do
+        @passenger = @dispatcher.find_passenger(54)
+      end
```

Comment:
Good work getting these edge cases. Organizationally, I think they belong under `passenger_spec` and `driver_spec` (because they test methods on those classes).